app/routes.py

```diff
@@ -1,2 +1,179 @@
+        task.completed_at = form_data["completed_at"]
+        db.session.commit()
+
+        return jsonify({"task":task.to_dict()}), 200
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if task == None:
+        return make_response(), 404
+
+    task.completed_at = datetime.datetime.now()
+    db.session.commit()
+
+    url = "https://slack.com/api/chat.postMessage"
+    slack_headers = {'Authorization': "SLACK_BOT_TOKEN"}
+    slack_params = {'channel':'task-notifications',
+            'text': f'Someone just completed the task {task.description}',
+            'format': "json"}
+    requests.post(url, headers=slack_headers, params=slack_params)
+
+    return jsonify({"task":task.to_dict()}), 200
+
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = Task.query.get(task_id)
+
+    if task == None:
+        return make_response(), 404
+
+    task.completed_at = None
+    db.session.commit()
+
+    return jsonify({"task":task.to_dict()}), 200
+
+@goal_bp.route("", methods=["POST", "GET"])
+def goals():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body:
+            return jsonify({"details": "Invalid data"}), 400
+        else:
+            new_goal = Goal(title = request_body["title"],)
+            db.session.add(new_goal)
+            db.session.commit()
+
+            return jsonify({"goal":new_goal.to_dict()}), 201
+
+    elif request.method == "GET":
+        goals = Goal.query.all()
+        goals_response = []
+        for goal in goals:
+            goals_response.append(goal.to_dict())
+
+        return jsonify(goals_response), 200
+
+@goal_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def get_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if goal == None:
+        return make_response(), 404
+    
+    elif request.method == "GET":
+        return jsonify({"goal":goal.to_dict()}), 200
+
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+        return jsonify({"details": f'Goal {goal.goal_id} "{goal.title}" successfully deleted'}),200
+
+    elif request.method == "PUT":
+        form_data = request.get_json()
+
+        goal.title = form_data["title"]
+        db.session.commit()
+
+        return jsonify({"goal":goal.to_dict()}), 200
+
+
+@goal_bp.route("/<goal_id>/tasks", methods=["GET", "POST"])
+def tasks_to_goals(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if goal == None:
+        return make_response(), 404
+    
+    elif request.method == "POST":
+        form_data = request.get_json() 
+        task_list = form_data["task_ids"]
+
+        for task in task_list:
+            task = Task.query.get(task)
+            task.assoc_goal = goal.goal_id
+            
+        db.session.commit()
```

Comment:
Good job remembering to commit here! Many people forgot since there isn't actually a test to verify the tasks are saved in the db.